src/System.Text.Json/src/System/Text/Json/JsonDocument.CustomDb.cs

```diff
@@ -0,0 +1,229 @@
+                // int estimatedTokens = payloadLength / 12
+                // now acknowledge that the number of bytes we need per token is 12.
+                // So that's just the payload length.
+                //
+                // Add one token's worth of data just because.
+                int initialSize = DbRow.Size + payloadLength;
+
+                // Stick with ArrayPool's rent/return range if it looks feasible.
+                // If it's wrong, we'll just grow and copy as we would if the tokens
+                // were more frequent anyways.
+                const int oneMegabyte = 1024 * 1024;
+                
+                if (initialSize > oneMegabyte && initialSize <= 4 * oneMegabyte)
+                {
+                    initialSize = oneMegabyte;
+                }
+
+                _rentedBuffer = ArrayPool<byte>.Shared.Rent(initialSize);
+                Length = 0;
+            }
+
+            public void Dispose()
+            {
+                if (_rentedBuffer == null)
+                {
+                    return;
+                }
+
+                ArrayPool<byte>.Shared.Return(_rentedBuffer);
+                _rentedBuffer = null;
+                Length = 0;
+            }
+
+            internal void TrimExcess()
+            {
+                if (Length <= _rentedBuffer.Length / 2)
+                {
+                    byte[] newRent = ArrayPool<byte>.Shared.Rent(Length);
+                    byte[] returnBuf = newRent;
+
+                    if (newRent.Length < _rentedBuffer.Length)
+                    {
+                        Buffer.BlockCopy(_rentedBuffer, 0, newRent, 0, Length);
+                        returnBuf = _rentedBuffer;
+                        _rentedBuffer = newRent;
+                    }
+
+                    ArrayPool<byte>.Shared.Return(returnBuf);
+                }
+            }
+
+            internal void Append(JsonTokenType tokenType, int startLocation, int length)
+            {
+                // StartArray or StartObject should have length -1, otherwise the length should not be -1.
+                Debug.Assert(
+                    (tokenType == JsonTokenType.StartArray || tokenType == JsonTokenType.StartObject) ==
+                    (length == DbRow.UnknownSize));
+
+                if (Length >= _rentedBuffer.Length - DbRow.Size)
+                {
+                    Enlarge();
+                }
+
+                DbRow row = new DbRow(tokenType, startLocation, length);
+                MemoryMarshal.Write(_rentedBuffer.AsSpan(Length), ref row);
+                Length += DbRow.Size;
+            }
+
+            private void Enlarge()
+            {
+                int size = _rentedBuffer.Length * 2;
+                byte[] newArray = ArrayPool<byte>.Shared.Rent(size);
+                Buffer.BlockCopy(_rentedBuffer, 0, newArray, 0, Length);
+                ArrayPool<byte>.Shared.Return(_rentedBuffer);
```

Comment:
See my other comment about Return'ing.  My preference would be to ensure that the field storing the buffer is cleared/replaced prior to returning the buffer, just to help with any potential reliability issues that could result in returning the buffer twice.  It's super unlikely, of course.